ipykernel/subshell_manager.py

```diff
@@ -0,0 +1,280 @@
+"""Manager of subshells in a kernel."""
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
+from __future__ import annotations
+
+import typing as t
+import uuid
+from dataclasses import dataclass
+from threading import Lock, current_thread, main_thread
+
+import zmq
+import zmq.asyncio
+from anyio import create_memory_object_stream, create_task_group
+
+from .subshell import SubshellThread
+
+
+@dataclass
+class Subshell:
+    thread: SubshellThread
+    shell_channel_socket: zmq.asyncio.Socket
+
+
+class SubshellManager:
+    """A manager of subshells.
+
+    Controls the lifetimes of subshell threads and their associated ZMQ sockets.
+    Runs mostly in the shell channel thread.
+
+    Care needed with threadsafe access here.  All write access to the cache occurs in
+    the shell channel thread so there is only ever one write access at any one time.
+    Reading of cache information can be performed by other threads, so all reads are
+    protected by a lock so that they are atomic.
+
+    Sending reply messages via the shell_socket is wrapped by another lock to protect
+    against multiple subshells attempting to send at the same time.
+    """
+
+    def __init__(self, context: zmq.asyncio.Context, shell_socket: zmq.asyncio.Socket):
+        assert current_thread() == main_thread()
+
+        self._context: zmq.asyncio.Context = context
+        self._shell_socket = shell_socket
+        self._cache: dict[str, Subshell] = {}
+        self._lock_cache = Lock()
+        self._lock_shell_socket = Lock()
+
+        # Inproc pair sockets for control channel and main shell (parent subshell).
+        # Each inproc pair has a "shell_channel" socket used in the shell channel
+        # thread, and an "other" socket used in the other thread.
+        self._control_shell_channel_socket = self._create_inproc_pair_socket("control", True)
+        self._control_other_socket = self._create_inproc_pair_socket("control", False)
+        self._parent_shell_channel_socket = self._create_inproc_pair_socket(None, True)
+        self._parent_other_socket = self._create_inproc_pair_socket(None, False)
+
+        # anyio memory object stream for async queue-like communication between tasks.
+        # Used by _create_subshell to tell listen_from_subshells to spawn a new task.
+        self._send_stream, self._receive_stream = create_memory_object_stream[str]()
+
+    def close(self) -> None:
+        """Stop all subshells and close all resources."""
+        assert current_thread().name == "Shell channel"
```

Comment:
```suggestion
        assert current_thread().name == SHELL_CHANNEL_THREAD_NAME
```

Maybe ? May need an import, and may trigger circular import, so I understand if not.

Reply:
I have moved the definitions of `CONTROL_THREAD_NAME` and `SHELL_CHANNEL_THREAD_NAME` to the `BaseThread` file as this avoids the circular imports. It makes sense anyway, thread names are ideally unique so if they are defined together you can see what the other ones are.

Reply:
Fixed in 9109cc8.